autopush/router/adm.py

```diff
@@ -0,0 +1,206 @@
+"""ADM Router"""
+import time
+import requests
+
+from typing import Any  # noqa
+
+from requests.exceptions import ConnectionError
+from twisted.internet.threads import deferToThread
+from twisted.logger import Logger
+
+from autopush.exceptions import RouterException
+from autopush.metrics import make_tags
+from autopush.router.interface import RouterResponse
+from autopush.types import JSONDict  # noqa
+
+
+class ADMAuthError(Exception):
+    pass
+
+
+class ADMClient(object):
+    def __init__(self,
+                 credentials=None,
+                 logger=None,
+                 metrics=None,
+                 endpoint="api.amazon.com",
+                 **options
+                 ):
+
+        self._client_id = credentials["client_id"]
+        self._client_secret = credentials["client_secret"]
+        self._token_exp = 0
+        self._auth_token = None
+        self._aws_host = endpoint
+        self._logger = logger
+        self._metrics = metrics
+        self._request = requests
+
+    def refresh_key(self):
+        url = "https://{}/auth/O2/token".format(self._aws_host)
+        if self._auth_token is None or self._token_exp < time.time():
+            body = dict(
+                grant_type="client_credentials",
+                scope="messaging:push",
+                client_id=self._client_id,
+                client_secret=self._client_secret
+            )
+            headers = {
+                "content-type": "application/x-www-form-urlencoded"
+            }
+            resp = self._request.post(url, data=body, headers=headers)
```

Comment:
we definitely want to specify some kind of timeout to requests calls, GCM bridge uses timeout=10. Then you have to catch the Timeout exception somewhere (probably _route)

Reply:
agreed.